src/rust/inetstack/protocols/layer4/tcp/established/receiver.rs

```diff
@@ -274,60 +186,166 @@ impl Receiver {
             warn!("Got packet with URG bit set!");
         }
 
-        let has_data: bool = if data.len() > 0 {
-            Self::process_data(cb, layer3_endpoint, data, seg_start, seg_end, seg_len)?;
-            true
-        } else {
-            false
-        };
+        // Store whether the packet has data here because processing it will consume the DemiBuffer.
+        let has_data: bool = data.len() > 0;
+        Self::process_data_if_any(cb, layer3_endpoint, data, seg_start, seg_end, seg_len)?;
```

Comment:
Can this call be avoided by checking `has_data` above?

Reply:
No, the data gets adjusted by the check_segment_in_window function. Given that each processing function might modify the buf, I thought doing a late binding was the best idea.

Reply:
The whole body of `process_data_if_any` is wrapped in the `data.len() > 0` check. So it will be enough to just check for `has_data` before calling it instead. Also, this function is called just from here as of now.

Reply:
oh, I'm storing the data for later in the function below process_data which takes ownership of data, so I can't check it later.

Reply:
Thanks for changing. The code is more readable now.